packages/@expo/cli/src/start/server/middleware/__tests__/CorsMiddleware-test.ts

```diff
@@ -0,0 +1,110 @@
+
+  beforeEach(() => {
+    resHeaders = {};
+    res = {
+      setHeader: (key: string, value: string) => {
+        resHeaders[key] = value;
+      },
+    } as unknown as ServerResponse;
+    next = jest.fn();
+    jest.resetAllMocks();
+  });
+
+  it('should pass through requests without origin', () => {
+    middleware(asRequest({ url: 'http://localhost:8081/', headers: {} }), res, next);
+    expect(resHeaders['Access-Control-Allow-Origin']).toBeUndefined();
+    expect(next).toHaveBeenCalled();
+  });
+
+  it('should allow CORS from localhost', () => {
+    const origin = 'http://localhost:8082/';
+    middleware(asRequest({ url: 'http://localhost:8081/', headers: { origin } }), res, next);
+    expect(resHeaders['Access-Control-Allow-Origin']).toBe(origin);
+    expect(next).toHaveBeenCalled();
+  });
+
+  it('should allow CORS from devtools://devtools', () => {
+    const origin = 'devtools://devtools';
+    middleware(asRequest({ url: 'http://localhost:8081/', headers: { origin } }), res, next);
+    expect(resHeaders['Access-Control-Allow-Origin']).toBe(origin);
+    expect(next).toHaveBeenCalled();
+  });
+
+  it('should allow CORS from https://chrome-devtools-frontend.appspot.com/', () => {
+    const origin = 'https://chrome-devtools-frontend.appspot.com/';
+    middleware(asRequest({ url: 'http://localhost:8081/', headers: { origin } }), res, next);
+    expect(resHeaders['Access-Control-Allow-Origin']).toBe(origin);
+    expect(next).toHaveBeenCalled();
+  });
+
+  it(`should allow CORS from expo-router's origin`, () => {
+    const origin = 'https://example.org/';
+    const middleware = createCorsMiddleware({ ...config, extra: { router: { origin } } });
+    middleware(asRequest({ url: 'http://localhost:8081/', headers: { origin } }), res, next);
+    expect(resHeaders['Access-Control-Allow-Origin']).toBe(origin);
+    expect(next).toHaveBeenCalled();
+  });
+
+  it(`should allow CORS from expo-router's headOrigin`, () => {
+    const headOrigin = 'https://example.org/';
+    const middleware = createCorsMiddleware({ ...config, extra: { router: { headOrigin } } });
+    middleware(
+      asRequest({ url: 'http://localhost:8081/', headers: { origin: headOrigin } }),
+      res,
+      next
+    );
+    expect(resHeaders['Access-Control-Allow-Origin']).toBe(headOrigin);
+    expect(next).toHaveBeenCalled();
+  });
+
+  it('should prevent metro reset the hardcoded CORS header', () => {
+    const origin = 'http://localhost:8082/';
+    middleware(
+      asRequest({ url: 'http://localhost:8081/index.map', headers: { origin } }),
+      res,
+      next
+    );
+    // Simulate metro to reset the hardcoded CORS header
+    res.setHeader('Access-Control-Allow-Origin', 'devtools://devtools');
+
+    expect(resHeaders['Access-Control-Allow-Origin']).toBe(origin);
+    expect(next).toHaveBeenCalled();
+  });
+
+  it('should show explicit error from disallowed CORS', () => {
+    const origin = 'https://example.org/';
```

Comment:
Probably want a test to ensure a complex URL isn't blocked, e.g. `https://example.org/foo/bar?alpha=beta#gamma`